Desktop Add colour hints to the local search bar.

mic704b opened the pull request:
Implements enhancement request #1861.

When no search results exist for the ongoing search, and the ongoing search is non-trivial, then the background colour of the search box is changed to warning colour, and so makes it obvious to the user.

laurent22 commented on ElectronClient/app/gui/NoteSearchBar.jsx:
```diff
@@ -113,11 +114,33 @@ class NoteSearchBarComponent extends React.Component {
+				this.state.backgroundColor = theme.backgroundColor;
+			} else {
+				this.state.backgroundColor = theme.warningBackgroundColor;
+			}
```
Please use `setState()` to set the state, so as to trigger a re-render.

mic704b replied:
State update eliminated in response to your other comments.

laurent22 commented:
Thanks for the change. In general we shouldn't change the state in `render()` as that would create a rendering loop.

But I think in your case the background colour is simply derived from the state and props so you don't need to create a state variable. Just create `let backgroundColor = "some default"` and then change it based on the state and props. I hope that makes sense and that I understood the logic correctly. If not please let me know.

mic704b commented:
> Thanks for the change. In general we shouldn't change the state in `render()` as that would create a rendering loop.

Sorry, I missed that.  I've removed it now.

> But I think in your case the background colour is simply derived from the state and props so you don't need to create a state variable. Just create `let backgroundColor = "some default"` and then change it based on the state and props.

Yes, that was my first instinct too.

The issue is that there are two renders of `SearchBox` upon each key press. The first is before the new results are known, the second is after.  As a result, the search box flashes in some cases eg when the first character is entered. It is only a small issue but not nice user experience.

So we need to keep track of the previous complete search information so that on the first render we don't change the background (in response to the new query and old result), and then update it only when we know the new result as well (new query and new result).

I updated it now without using the state update. Is this better?

laurent22 commented:
Thanks for clarifying @mic704b. I'll need to review but now that you've added backgroundColor as a class member, I feel it's actually part of the state and that's where it should go, so we're a bit back at the beginning. Perhaps the issue is that the backgroundColor should be set on componentDidUpdate, but not sure as I don't quite understand the logic anymore. I'll need to review this more in details.

mic704b commented:
Ok sounds good. I might have another look if I get a chance, but otherwise I'll await your review. Thanks

mic704b commented:
I think it's improved now.  

I've also added text to show the number of matches, and fade out prev/next buttons when no matches  exist.

mic704b commented:
I've broken this branch while rebasing, so will close this pull request and create a new one (#2329).
